File: plugins/dotnet/robocode.dotnet.api.net/src/BattleResults.cs
using System;
using System.IO;
using net.sf.robocode.serialization;

namespace robocode
{
    public class BattleResults : Object, IComparable
    {
        protected internal const long serialVersionUID = 1L;
        protected internal double bulletDamage;
        protected internal double bulletDamageBonus;
        protected internal int firsts;
        protected internal double lastSurvivorBonus;
        protected internal double ramDamage;
        protected internal double ramDamageBonus;
        protected internal int rank;
        protected internal double score;
        protected internal int seconds;
        protected internal double survival;
        protected internal string teamLeaderName;
        protected internal int thirds;

        public BattleResults(string teamLeaderName, int rank, double score, double survival, double lastSurvivorBonus,
                             double bulletDamage, double bulletDamageBonus, double ramDamage, double ramDamageBonus,
                             int firsts, int seconds, int thirds)
        {
            this.teamLeaderName = teamLeaderName;
            this.rank = rank;
            this.score = score;
            this.survival = survival;
            this.lastSurvivorBonus = lastSurvivorBonus;
            this.bulletDamage = bulletDamage;
            this.bulletDamageBonus = bulletDamageBonus;
            this.ramDamage = ramDamage;
            this.ramDamageBonus = ramDamageBonus;
            this.firsts = firsts;
            this.seconds = seconds;
            this.thirds = thirds;
        }

        #region IComparable Members

        int IComparable.CompareTo(object obj1)
        {
            if (obj1 is BattleResults)
            {
                return score.CompareTo(((BattleResults) obj1).score);
            }
            return -1;
        }

        #endregion

        internal static ISerializableHelper createHiddenSerializer()
        {
            return new SerializableHelper(null);
        }

        public virtual int getBulletDamage()
        {
            return (int) (bulletDamage + 0.5);
        }

        public virtual int getBulletDamageBonus()
        {
            return (int) (bulletDamageBonus + 0.5);
        }

        public virtual int getFirsts()
        {
            return firsts;
        }

        public virtual int getLastSurvivorBonus()
        {
            return (int) (lastSurvivorBonus + 0.5);
        }

        public virtual int getRamDamage()
        {
            return (int) (ramDamage + 0.5);
        }

        public virtual int getRamDamageBonus()
        {
            return (int) (ramDamageBonus + 0.5);
        }

        public virtual int getRank()
        {
            return rank;
        }

        public virtual int getScore()
        {
            return (int) (score + 0.5);
        }

        public virtual int getSeconds()
        {
            return seconds;
        }

        public virtual int getSurvival()
        {
            return (int) (survival + 0.5);
        }

        public virtual string getTeamLeaderName()
        {
            return teamLeaderName;
        }

        public virtual int getThirds()
        {
            return thirds;
        }

        /*public static implicit operator Serializable(BattleResults results1)
        {
            Serializable serializable;
            serializable.__<ref> = results1;
            return serializable;
        }*/

        #region Nested type: SerializableHelper

        internal sealed class SerializableHelper : ISerializableHelper
        {
            private SerializableHelper()
            {
            }

            internal SerializableHelper(BattleResults a1) : this()
            {
            }

            #region ISerializableHelper Members

            public object deserialize(RbSerializer serializer, BinaryReader br)
            {
                string teamLeaderName = serializer.deserializeString(br);
                int rank = br.ReadInt32();
                double score = br.ReadDouble();
                double survival = br.ReadDouble();
                double lastSurvivorBonus = br.ReadDouble();
                double bulletDamage = br.ReadDouble();
                double bulletDamageBonus = br.ReadDouble();
                double ramDamage = br.ReadDouble();
                double ramDamageBonus = br.ReadDouble();
                int firsts = br.ReadInt32();
                int seconds = br.ReadInt32();
                return new BattleResults(teamLeaderName, rank, score, survival, lastSurvivorBonus, bulletDamage,
                                         bulletDamageBonus, ramDamage, ramDamageBonus, firsts, seconds, br.ReadInt32());
            }

            public void serialize(RbSerializer serializer, BinaryWriter bw, object obj)
            {
                var results = (BattleResults) obj;
                serializer.serialize(bw, results.teamLeaderName);
                serializer.serialize(bw, results.rank);
                serializer.serialize(bw, results.score);
                serializer.serialize(bw, results.survival);
                serializer.serialize(bw, results.lastSurvivorBonus);
                serializer.serialize(bw, results.bulletDamage);
                serializer.serialize(bw, results.bulletDamageBonus);
                serializer.serialize(bw, results.ramDamage);
                serializer.serialize(bw, results.ramDamageBonus);
                serializer.serialize(bw, results.firsts);
                serializer.serialize(bw, results.seconds);
                serializer.serialize(bw, results.thirds);
            }

            public int sizeOf(RbSerializer serializer1, object obj1)
            {
                var results = (BattleResults) obj1;
                return (((1 + serializer1.sizeOf(results.teamLeaderName)) + 0x10) + 0x38);
            }

            #endregion
        }

        #endregion
    }
}

File: source_code/robocode.api/src/main/java/robocode/BattleResults.java
/*******************************************************************************
 * Copyright (c) 2001, 2008 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Common Public License v1.0
 * which accompanies this distribution, and is available at
 * http://robocode.sourceforge.net/license/cpl-v10.html
 *
 * Contributors:
 *     Pavel Savara
 *     - Initial implementation
 *     Flemming N. Larsen
 *     - Javadocs
 *     Joshua Galecki
 *     - Made into an interface, combined scores into lists
 *******************************************************************************/

package robocode;


import java.nio.ByteBuffer;
import java.util.List;

import net.sf.robocode.serialization.ISerializableHelper;
import net.sf.robocode.serialization.RbSerializer;


/**
 * Contains the battle results returned by {@link BattleEndedEvent#getResults()}
 * when a battle has ended.
 *
 * @author Pavel Savara (original)
 * @author Joshua Galecki (contributor)
 * 
 * @see BattleEndedEvent#getResults()
 * @see Robot#onBattleEnded(BattleEndedEvent)
 * @since 1.6.1
 */
public class BattleResults implements IBattleResults {

	/**
	 * 
	 */
	protected static final long serialVersionUID = 1L;
	
	protected String teamName;
	protected int rank;
	protected double combinedScore;
	protected List<Double> scores;
	protected List<String> scoreNames;
	protected int firsts;
	protected int seconds;
	protected int thirds;
	
	public BattleResults() {}
	
	public BattleResults(String teamName, int rank, double combinedScore, List<Double> scores,
			List<String> scoreNames, int firsts, int seconds, int thirds) {
		this.teamName = teamName;
		this.rank = rank;
		this.combinedScore = combinedScore;
		this.scores = scores;
		this.scoreNames = scoreNames;
		this.firsts = firsts;
		this.seconds = seconds;
		this.thirds = thirds;
	}
	
	/**
	 * {@inheritDoc}
	 */
	public int compareTo(IBattleResults o) {
		if (combinedScore == o.getCombinedScore()) {
			return 0;
		} else if (combinedScore > o.getCombinedScore()) {
			return 1; 
		} else {
			return -1;
		}
	}

	/**
	 * Returns the name of the team or the name of the
	 * robot if the robot is not participating in a team.
	 *
	 * @return the name of the team or the name of the robot.
	 */
	public String getTeamName() {
		return teamName;
	}

	/**
	 * Returns the rank of this robot in the battle results.
	 *
	 * @return the rank of this robot in the battle results.
	 */
	public int getRank() {
		return rank;
	}

	/**
	 * Returns the total score of this robot in the battle.
	 *
	 * @return the total score of this robot in the battle.
	 */
	public double getCombinedScore() {
		return combinedScore;
	}
	
	/**
	 * Returns a list of the scores of the robot.
	 * 
	 * @return a list of the scores of the robot.
	 */
	public List<Double> getScores() {
		return scores;
	}
	
	/**
	 * Returns a list of the different categories of scores for the robot.
	 * 
	 * @return a list of the different categories of scores for the robot.
	 */
	public List<String> getScoreNames() {
		return scoreNames;
	}

	/**
	 * Returns the number of rounds this robot placed first in the battle.
	 *
	 * @return the number of rounds this robot placed first in the battle.
	 */
	public int getFirsts() {
		return firsts;
	}

	/**
	 * Returns the number of rounds this robot placed second in the battle.
	 *
	 * @return the number of rounds this robot placed second in the battle.
	 */
	public int getSeconds() {
		return seconds;
	}
	
	/**
	 * Returns the number of rounds this robot placed third in the battle.
	 *
	 * @return the number of rounds this robot placed third in the battle.
	 */
	public int getThirds() {
		return thirds;
	}
	
	/**
	 * {@inheritDoc}
	 */
	public int compareTo(BattleResults o) {
		return ((Double) combinedScore).compareTo(o.getCombinedScore());
	}

	static ISerializableHelper createHiddenSerializer() {
		return new SerializableHelper();
	}

	private static class SerializableHelper implements ISerializableHelper {
		public int sizeOf(RbSerializer serializer, Object object) {
			BattleResults obj = (BattleResults) object;

			return RbSerializer.SIZEOF_TYPEINFO + serializer.sizeOf(obj.getTeamName()) + 4 * RbSerializer.SIZEOF_INT
					+ 7 * RbSerializer.SIZEOF_DOUBLE;
		}

		public void serialize(RbSerializer serializer, ByteBuffer buffer, Object object) {
			BattleResults obj = (BattleResults) object;

			double scoresArray[] = new double[obj.getScores().size()];

			for (int scoreIndex = 0; scoreIndex < obj.getScores().size(); scoreIndex++) {
				scoresArray[scoreIndex] = obj.getScores().get(scoreIndex);
			}
			
			// TODO: test, ensure that a \n cannot be inserted into a score's name
			int namesLength = 0;

			for (String name : obj.getScoreNames()) {
				namesLength += name.length() + 1;
			}
			char nameArray[] = new char[namesLength];
			int arrayIndex = 0;

			for (String name : obj.getScoreNames()) {
				for (int charIndex = 0; charIndex < name.length(); charIndex++) {
					nameArray[arrayIndex] = name.charAt(charIndex);
					arrayIndex++;
				}
				nameArray[arrayIndex] = '\n';
				arrayIndex++;
			}
				
			serializer.serialize(buffer, obj.getTeamName());
			serializer.serialize(buffer, obj.rank);
			serializer.serialize(buffer, obj.getCombinedScore());
			serializer.serialize(buffer, scoresArray);
			serializer.serialize(buffer, nameArray);
			serializer.serialize(buffer, obj.firsts);
			serializer.serialize(buffer, obj.seconds);
			serializer.serialize(buffer, obj.thirds);
		}

		public Object deserialize(RbSerializer serializer, ByteBuffer buffer) {
			String teamName = serializer.deserializeString(buffer);
			int rank = buffer.getInt();
			double score = buffer.getDouble();

			// TODO: unserializing the buffer.
			// double survival = buffer.getDouble();
			// double lastSurvivorBonus = buffer.getDouble();
			// double bulletDamage = buffer.getDouble();
			// double bulletDamageBonus = buffer.getDouble();
			// double ramDamage = buffer.getDouble();
			// double ramDamageBonus = buffer.getDouble();
			// int firsts = buffer.getInt();
			// int seconds = buffer.getInt();
			// int thirds = buffer.getInt();
			//
			// return new BattleResults(teamLeaderName, rank, score, survival, lastSurvivorBonus, bulletDamage,
			// bulletDamageBonus, ramDamage, ramDamageBonus, firsts, seconds, thirds);
			return new BattleResults(teamName, rank, score, null, null, 0, 0, 0);
		}
	}
}
